feat: add ranking table

Fruktus opened the pull request:
Introduces new ranking table, which is used to provide the current ranking when players join lobby (should slightly help with CPU since it doesn't require recalculation) and will help us retain old leaderboards when the ranking calculation will change

kjarosh commented on server/src/QRServer/db/connector.py:
```diff
@@ -288,13 +303,70 @@ async def get_ranking(self, start_date: datetime, end_date: datetime, ranked_onl
+        await self.conn.commit()
+
+    async def _generate_ranking(self, start_date: datetime, end_date: datetime, ranked_only=True,
+                                include_void=False) -> List[RankingEntry]:
```
Why accept those values as parameters when you can just get them here?

Fruktus replied:
what do you mean? as in generate datetime?

Fruktus replied:
Use config values for `ranked_only` and `include_void`

kjarosh commented on server/tests/test_utils.py:
```diff
@@ -8,15 +8,56 @@ class MakeMonthDatesTest(unittest.TestCase):
+            ]
+        )
+
+        self.assertEqual(
```
Split it into different tests? + No test for multiple years

Fruktus replied:
As in separate test for generating ranking and adding match? Or how?

Fruktus commented on server/src/QRServer/db/connector.py:
```diff
@@ -288,13 +303,70 @@ async def get_ranking(self, start_date: datetime, end_date: datetime, ranked_onl
             " inner join matches m on (u.id = m.winner_id or u.id = m.loser_id)"
             " where m.started_at >= ?"
-            " and m.finished_at < ?"
+            " and m.finished_at <= ?"
```
I'm not sure I follow

Fruktus replied:
I'll rewrite it to only filter by end time

kjarosh replied:
The new logic also does not make sense, it will duplicate matches

Fruktus replied:
Right, corrected

Fruktus commented:
@kjarosh Manual testing report:
1. Downloaded a copy of db, verified its contents (matches present, no `ranking` table)
2. Ran the server (to apply migration) and stopped, verified with sqlite - `ranking` table created and populated
3. Ran the server again, verified that the leaderboards appear correctly (they seem to have the same content, but due to our change in ranking they differ from before - although the current ones changed also)
4. Stopped the server, removed all entries from `matches`, restarted server - ranking still works, no changes (so they are correctly persisted)

Fruktus commented:
> @kjarosh Manual testing report:
> 
>     1. Downloaded a copy of db, verified its contents (matches present, no `ranking` table)
> 
>     2. Ran the server (to apply migration) and stopped, verified with sqlite - `ranking` table created and populated
> 
>     3. Ran the server again, verified that the leaderboards appear correctly (they seem to have the same content, but due to our change in ranking they differ from before - although the current ones changed also)
> 
>     4. Stopped the server, removed all entries from `matches`, restarted server - ranking still works, no changes (so they are correctly persisted)

@kjarosh followup - I also tested adding the match (set up member vs member game) and it did correctly add, with exception of not clearing previous entries, since I dropped all the matches from db. IMHO it is not a problem since this occured due to me manually removing records but something to be aware of. Overall it seems to work as expected

kjarosh commented on server/src/QRServer/db/migrations.py:
```diff
@@ -54,6 +60,39 @@ async def execute_migrations(c):
+            for date in dates_range:
+                start_date, end_date = utils.make_month_dates(date.month, date.year)
+
+                ranking_entries = await dbconnector._generate_ranking(start_date, end_date)
```
Let's copy the query here. That will prevent errors in the future if we decide to change the behavior of this method. That will also not require passing dbconnector

Fruktus replied:
Done. Repeated the above tests (all of them), still works as expected

kjarosh replied:
Where did you copy the query? I can't see it

Fruktus replied:
![image](https://github.com/user-attachments/assets/be163a21-abb4-4acd-b1bd-15fe08da41aa)

kjarosh commented on server/src/QRServer/db/connector.py:
```diff
@@ -187,6 +189,19 @@ async def add_match_result(self, match_result: DbMatchReport):
+        new_ranking = await self._generate_ranking(
+            start_date, end_date)
+
+        if (not ranked_only or match_result.is_ranked) and (not include_void or not match_result.is_void):
```
This condition seems to be wrong, it should be covered by tests

kjarosh replied:
BTW do we need this condition? We're basically duplication logic at this point

Fruktus commented on server/src/QRServer/db/migrations.py:
```diff
@@ -64,3 +39,148 @@ async def _select_version(c):
+
+
+async def _migration_upgrade_to_v2(c, _config):
+    await c.execute("alt\er table users add column comment varchar")
```
```suggestion
    await c.execute("alter table users add column comment varchar")
```
